modules/nextflow/src/main/groovy/nextflow/script/params/TupleInParam.groovy

```diff
@@ -79,6 +80,9 @@ class TupleInParam extends BaseInParam {
             else if( item instanceof TokenStdinCall ) {
                 newItem(StdInParam)
             }
+            else if( item instanceof TokenCmdCall ) {
+                throw new IllegalArgumentException('Command input declaration is not supported')
+            }
```

Comment:
I keep wondering what it would mean to have a "command input"... here are some ideas

If we ever decide to do more fine-grained resource optimization based on task inputs, I would like to be able to specify custom metrics based on the input files. We can already get the file size, but file size isn't always the most salient metric for resource usage. See [here](https://github.com/SystemsGenetics/KINC-nf/blob/77545302244c50278b322c8150ca68c6f4a14ee9/main.nf#L179-L181) how I specified custom metrics for my PhD research.

Perhaps a "command input" could serve this purpose, but the challenge is that it needs to run before the task is launched, if it is to be used to influence process directives like cpus, memory, etc

Here is a sketch based on the example I linked:
```groovy
def predict_memory(n_rows, n_cols) {
  // query prediction model
}

process IMPORT_EMX {
  memory { predict_memory(n_rows, n_cols) }

  input:
  path 'input.txt'
  cmd 'tail -n +1 input.txt | wc -l', name: 'n_rows'
  cmd 'head -n +1 input.txt | wc -w', name: 'n_cols'

  // ...
}
```

This syntax is confusing because the command inputs aren't really process inputs, i.e. the user doesn't need to supply input channels for them. Seems like the command input should really just be a function that has access to the task inputs:

```groovy
process IMPORT_EMX {
  memory {
    def n_rows = cmd('tail -n +1 input.txt | wc -l')
    def n_cols = cmd('head -n +1 input.txt | wc -w')
    predict_memory(n_rows, n_cols)
  }

  input:
  path 'input.txt'

  // ...
}
```

This is still dubious because directives are evaluated before the task environment is created and input files staged. We might just have to limit ourselves to task metadata:

```groovy
  memory {
    predict_memory( task.inputs['input.txt'].size() )
  }
```

This is all outside the scope of this PR, just leaving my notes here for future reference.

Reply:
Shall you fork out a new issue out of this interesting note?! :-)

Reply:
See #4545

Reply:
thank you!